### PR TITLE
drivers: ethernet: stm32: Enable multicast address filter

### PR DESCRIPTION
**Pull request by mrfuchs**
Use the MAC's hash table to filter incoming multicast frames that are
targeted to subscribed multicast groups and discard all others.

Signed-off-by: Markus Fuchs <markus.fuchs@ch.sauter-bc.com>

### CODE REVIEW

**Comment by Nukersson** [comment]
Please wait until today evening, I'll do the review. There are many places, needed to be changed.

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:147]:
```suggestion
#ifdef CONFIG_ETH_MULTICAST_FILTER
  	MACFilterConf.HashMulticast = ENABLE;
  	MACFilterConf.PassAllMulticast = DISABLE;
#else 
        MACFilterConf.HashMulticast = DISABLE;
  	MACFilterConf.PassAllMulticast = ENABLE;
#endif  /* CONFIG_ETH_MULTICAST_FILTER */
```

Please add CONFIG_ETH_MULTICAST_FILTER in KConfig also. It would be better to make it controlled by KConfig.

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:158]:
```suggestion
#ifndef CONFIG_ETH_MULTICAST_FILTER
               ETH_MULTICASTFRAMESFILTER_HASHTABLE |
#endif  /* CONFIG_ETH_MULTICAST_FILTER */
		 ETH_MULTICASTFRAMESFILTER_PERFECT);
```

Comment (Nukersson) [reply drivers/ethernet/eth_stm32_hal.c:158]: Must be guarded with CONFIG_ETH_MULTICAST_FILTER

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:161]:
```suggestion
#ifdef CONFIG_ETH_MULTICAST_FILTER
	/* enable multicast hash receive filter */
	tmp |= ETH_MULTICASTFRAMESFILTER_HASHTABLE;
#else
        /* enable receiving all multicast frames */
 	tmp |= ETH_MULTICASTFRAMESFILTER_NONE;
#endif  /* CONFIG_ETH_MULTICAST_FILTER */
```

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:790]:
```suggestion
```

Please use appropriate function from [byteorder.h](https://docs.zephyrproject.org/apidoc/latest/byteorder_8h.html) instead

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:899]:
```suggestion
#ifdef CONFIG_ETH_MULTICAST_FILTER
	net_if_mcast_mon_register(&mcast_monitor, iface, net_if_mcast_cb);
#endif  /* CONFIG_ETH_MULTICAST_FILTER */
```

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:791]:
```suggestion
#ifdef CONFIG_ETH_MULTICAST_FILTER
```

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:874]:
```suggestion
#endif  /* CONFIG_ETH_MULTICAST_FILTER */
```

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:872]:
```suggestion
#endif  /* CONFIG_SOC_SERIES_STM32H7X */
```

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:849]:
```suggestion
               LOG_DBG("???");
		return;
```
Please put meaningful debug message here.

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:812]:
```suggestion
	uint32_t i;
```
The used ´i´ variable is always positive here!

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:830]:
```suggestion
               LOG_DBG("???");
		return;
```

Please put meaningful debug message here!

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:843]:
```suggestion
		hash_table[hash_index % 32] |= (1 << (hash_index % 32));
```

Wouldn't be i more correct?

File: drivers/ethernet/eth_stm32_hal.c
Comment (Nukersson) [inline drivers/ethernet/eth_stm32_hal.c:862]:
```suggestion
		hash_table[hash_index % 32] |= (1 << (hash_index % 32));
```

Same as above...

**Review by KozhinovAlexander — CHANGES_REQUESTED** [review]
Thank you very much for your effort. Please accept the changes or give argumented point of your view to not accept them.

I would prefer to add `CONFIG_ETH_MULTICAST_FILTER` to stm32's KConfig. If you not agree please give me argumented answer. 

Please also describe tests you done and on which SoC. I'll try to test it at weekend with my HW.

**Comment by erwango** [comment]
@mrfuchs Do you plan to continue this PR ?

File: drivers/ethernet/eth_stm32_hal.c
Comment (jukkar) [inline drivers/ethernet/eth_stm32_hal.c:833]: Why do we loop through all multicast addresses here? This callback is called for those multicast addresses that we are joining, and it is possible that the array will contain addresses that are not joined and in that case that address should not be added to the hash. So it should be enough to just use the IP address given to this callback and use only that address when setting up the hash.

File: drivers/ethernet/eth_stm32_hal.c
Comment (jukkar) [inline drivers/ethernet/eth_stm32_hal.c:852]: Same thing here. We should remove this loop and only add the address set in the callback.

**Review by jukkar — CHANGES_REQUESTED** [review]

**Comment by cperera-aud** [comment]
Hi @mrfuchs, @erwango, @jukkar, @Nukersson,

We're interested in ensuring that this PR gets merged in and we're more than happy to continue the work that @mrfuchs left off and address review comments and also do testing etc.

Our current priority is to integrate https://github.com/zephyrproject-rtos/zephyr/pull/45770. With time we will look into getting the multicast hash filter going for the STM32H7 since its an absolute must for our application because we rely on a lot of multicast. Our devices have to be robust on networks without a managed switch and with large amounts of multicast (e.g. 60% of total available bandwidth). 

What I wasn't sure was given that the work was started by @mrfuchs, what is involved in for us to come in and continue the work. If it was within an organisation (e.g. company) then the process is quite clear. In the case of opensource I wasn't sure. I also wanted to ensure that handover happens quite cleanly and transparently.

**Comment by jukkar** [comment]
@cperera-aud you can take the code and enhance it where needed, then just re-submit it. Just remember to leave any existing copyright notices and signoff-by to existing code, and add your own copyrights/signoffs where applicable.

**Comment by erwango** [comment]
For reference: https://docs.zephyrproject.org/latest/contribute/guidelines.html#notes

**Comment by mrfuchs** [comment]
@cperera-aud, happy to hear you want to help getting this PR merged.

As @jukkar suggests, just take my code and enhance it where needed.

As for the remaining change requests, six of them suggest to introduce a new option `CONFIG_ETH_MULTICAST_FILTER` which - from my point of view - would classify this patch as an enhancement.
I always looked at these changes as a fix, see the message of the driver's second commit ad8f837 and the corresponding discussion in PR [#538](https://github.com/zephyrproject-rtos/zephyr/pull/538#issuecomment-312328421)).

The two requests to remove the loops over the multicast addresses in the `net_if_mcast_cb()` callback can't be right. Changing the code as suggested would mean multicast addresses can only be added, but never removed.

If you've more questions, I'm happy to help.

**Comment by cperera-aud** [comment]
Thanks @jukkar and @erwango for getting back and providing the information and links. Sorry it seems like I missed the info on contributing to existing PRs.

**Comment by cperera-aud** [comment]
Thanks @mrfuchs for replying and providing your thoughts. I will humbly mention that I will be unbiased in providing my thoughts while also acknowledging your hard work. 

For starters regarding the kconfig option `CONFIG_ETH_MULTICAST_FILTER` I don't have a strong opinion and I do see advantages on disadvantages on both sides of the argument. Looking at other equivalent ethernet drivers in zephyr/drivers/ethernet they don't seem to use this config option. Also having the multicast hash filter on all the time is certainly a good thing.  

On the other hand by having this config option allows the flexibility to turn off and turn on the feature at compile time. While I was working on our application, when the device was receiving low rate multicast e.g. mDNS, UDP/IP PTPV1 not having the multicast hash filter was not an issue. When our application started to support multicast at higher rates e.g. networked audio the need for the multicast hash filter was more apparent. Sorry I'm not sure in this instance who should have the final say on this matter any help to move this forward would be greatly appreciated. I understand that you mentioned then this PR should be an enhancement as opposed to a bug fix, then is there anything that has to be done to turn the PR into an enhancement?

In the case of the loop, again without being biased, I tend to agree with @jukkar on this one, because normally we should be adding one address at a time to the hash. Also I think we need to add some additional code for the case of removing entries from the multicast hash filter. We have a similar implementation for the STM32F4 series (proprietary implementation) but on a FreeRTOS/LWIP platform and in that case we add one multicast address at a time and remove one address at a time. e.g (just a guess). 

`hash_table[hash_index / 32] &= ~(1 << (hash_index % 32));`

Please let me know how to best move forward with the PR.

**Comment by mrfuchs** [comment]
> we add one multicast address at a time and remove one address at a time. e.g (just a guess).
> 
> `hash_table[hash_index / 32] &= ~(1 << (hash_index % 32));`

Wouldn't this remove multiple addresses from the hardware filter in case of a hash collision?

**Comment by cperera-aud** [comment]
> > we add one multicast address at a time and remove one address at a time. e.g (just a guess).
> > `hash_table[hash_index / 32] &= ~(1 << (hash_index % 32));`
> 
> Wouldn't this remove multiple addresses from the hardware filter in case of a hash collision?

Thanks @mrfuchs for getting back. So in your change by looping through all the multicast addresses regardless of them being IGMP joined or not was to avoid this scenario?

**Comment by cperera-aud** [comment]
Hi @mrfuchs I take back my comment regarding the method for updating the multicast hash filter for removing multicast addresses. It certainly is better to build the entire hash table by looping through the addresses that have joined. For adding a multicast address it really isn't necessary to rebuild the hash table.

**Pull request closed without merging**
